### ISSUE TITLE
nmap incorrectly detects TFTP port as open|filtered

### ISSUE THREAD

**Issue by guser99** [nmap/nmap#2406]
**Describe the bug**
 - In version 7.92, an open tftp port is not recognized as open.
>nmap -Pn -n -sU -p 69 10.129.223.3
```Starting Nmap 7.92 ( https://nmap.org ) at 2021-11-10 10:59 CET
Nmap scan report for 10.129.223.3
Host is up.

PORT   STATE         SERVICE
69/udp open|filtered tftp

Nmap done: 1 IP address (1 host up) scanned in 2.14 seconds
```

![wireshark_nmap_7 92](https://user-images.githubusercontent.com/91061336/141109549-8567f0da-568a-4219-97ab-167a4a9b3d93.png)
 - Note in the screenshot the column SPort and DPort.
 - In wireshark you can see the response of the TFTP server, but it comes from a different source port than 69, so nmap marks the port as open|filtered.

**To Reproduce**
 - Scan a TFTP Server with:
>nmap -Pn -n -sU -p 69 <TFTP-Server>

**Expected behavior**
 - The port is detected as open as in version 7.91.

>nmap -n -Pn -sU -p 69 10.129.223.3
```Host discovery disabled (-Pn). All addresses will be marked 'up' and scan times will be slower.
Starting Nmap 7.91 ( https://nmap.org ) at 2021-11-10 10:01 GMT
Nmap scan report for 10.129.223.3
Host is up (0.020s latency).

PORT   STATE SERVICE
69/udp open  tftp

Nmap done: 1 IP address (1 host up) scanned in 0.15 seconds
```
![wireshark_nmap_7 91](https://user-images.githubusercontent.com/91061336/141109528-5e5f8228-f5bd-4be7-bb61-3ad623e852cd.png)
 - Here you can see that the response of the TFTP server comes from the source port 69, so nmap marks the port as open.

**Version info (please complete the following information):**
 - OS:
1.
>Linux computer 5.10.0-kali9-amd64
2.
>Linux 5.13.19-2-MANJARO
 - Output of `nmap --version`:
1.
>Nmap version 7.91 ( https://nmap.org )
>Platform: x86_64-pc-linux-gnu
>Compiled with: liblua-5.3.3 openssl-1.1.1g libssh2-1.8.0 libz-1.2.11 libpcre-8.39 nmap-libpcap-1.7.3 nmap-libdnet-1.12 ipv6
>Compiled without:
>Available nsock engines: epoll poll select
2.
>Nmap version 7.92 ( https://nmap.org )
>Platform: x86_64-pc-linux-gnu
>Compiled with: liblua-5.3.6 openssl-1.1.1l libssh2-1.9.0 libz-1.2.11 libpcre-8.45 libpcap-1.10.1 nmap-libdnet-1.12 >ipv6
>Compiled without:
>Available nsock engines: epoll poll select

**Comment by dmiller-nmap** [comment]
This is an interesting complication of a change that was actually a fix for a different issue. Nmap 7.91 had a bug that caused UDP payloads to be corrupted, so it did not send the correct TFTP payload (note that the server says "Unknown operation code"). This particular TFTP daemon responds to bad messages using the same source port (69) as the destination, which is typical of most services. Good messages, like the one Nmap 7.92 sends, get the standard TFTP response from a different port, which is much harder to track.

All that to say, this is Nmap's expected behavior. However, I can think of 2 potential fixes:
1. Add port 69 to an existing payload that can get an error response from most TFTP services, so that Nmap can catch the port as open in any cases where errors are sent from port 69.
2. Add an exception in the scanning engine for port 69, allowing protocol responses to come from other ports as long as the destination port is correct. This is clunkier, but would find all conformant TFTP services.

Neither of these addresses the problem of identifying the TFTP server version, since that is done using Nsock with connected UDP sockets, and I do not think we can support receiving messages from arbitrary source ports. See #548 for a possible workaround there.

**Comment by nnposter** [comment]
> 1. Add port 69 to an existing payload that can get an error response from most TFTP services, so that Nmap can catch the port as open in any cases where errors are sent from port 69.

FWIW, I played with `atftpd` on Kali, trying to come up with a payload that would illicit a response from port 69. Nothing worked. I tried invalid opcodes, opcodes with non-zero high byte, short packets, etc. The daemon responded in all cases with an error message but its TID (effectively its source port) was always ephemeral, never 69.

**Comment by dmiller-nmap** [comment]
@nnposter Yeah, I'd guess most tftpd software will be like that, but this report is evidence that at least one doesn't, and I wouldn't be surprised if that's not the only one.

**Comment by dmiller-nmap** [comment]
I did some checking and a lot of software responds to our existing probe using port 69 as source port, including tftpd from Busybox, Brother printers, Cisco routers, Plan 9, Solarwinds, MS Windows Deployment Services, and Open TFTP Server. I got the list of systems from [Shodan](https://shodan.io/), so it may be skewed towards services that are easily discoverable in this way. About 30% of the systems I tested did not respond to our default probe, but adding a probe starting with 2 null bytes (DNSStatusRequest and a new probe, "\0\0\0\0", performed the same) improved the response rate to 90%. I will start by adding port 69 to the DNSStatusRequest payload, which I hazard to guess will fix this issue. Then I'll consider whether we can improve UDP scan generally by sending the "\0\0\0\0" payload *in addition to* the default empty payload for ports that do not have a payload defined, in order to catch more services. This will likely require further empirical testing to determine whether it's worth the effort.

**Comment by h4b4n3r0** [comment]
I also stumbled upon this issue when doing some tftp testing. I also experienced the same problems as guser99. Apparently it is still not fixed. Can we somehow support in this topic?
